community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/RotationState.java

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.neo4j.kernel.impl.store.kvstore;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.concurrent.locks.Lock;
+
+import org.neo4j.function.Consumer;
+import org.neo4j.helpers.Exceptions;
+import org.neo4j.helpers.Pair;
+import org.neo4j.kernel.impl.util.function.Optional;
+import org.neo4j.kernel.impl.util.function.Optionals;
+
+abstract class RotationState<Key> extends ProgressiveState<Key>
+{
+    abstract ProgressiveState<Key> rotate( RotationStrategy strategy, Consumer<Headers.Builder> headersUpdater )
+            throws IOException;
+
+    @Override
+    String stateName()
+    {
+        return "rotating";
+    }
+
+    static final class Rotation<Key> extends RotationState<Key>
+    {
+        private final ActiveState<Key> preState;
+        private final PrototypeState<Key> postState;
+        private final long threshold;
+
+        Rotation( ActiveState<Key> preState, PrototypeState<Key> postState, long version )
+        {
+            this.preState = preState;
+            this.postState = postState;
+            this.threshold = version;
+        }
+
+        ActiveState<Key> rotate( RotationStrategy strategy, Consumer<Headers.Builder> headersUpdater ) throws IOException
+        {
+            for ( long expected = preState.version() - preState.store.version(), sleep = 1;
+                  preState.applied() < expected; sleep = Math.min( sleep * 2, 100 ) )
+            {
+                try
+                {
+                    Thread.sleep( sleep );
+                }
+                catch ( InterruptedException e )
+                {
+                    throw Exceptions.withCause( new InterruptedIOException( "Rotation was interrupted." ), e );
+                }
+            }
```

Comment:
It would be nice to kill this live lock by registering a callback somehow on the state and being notified when all the transactions have been applied.